### PR TITLE
[BEAM-7587] Spark portable streaming

### PR DESCRIPTION
**Pull request by annaqin418**
These changes will allow streaming data on the portable Spark runner. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by annaqin418** [comment]
R: @ibzib 
Creating this PR for now to track changes as I go. I set up some of the new interfaces, could you check if they look reasonable? I currently have streaming context as a subclass of the original spark translation context, but thinking some more I'm not sure if that totally works, and it might be better to have them both implementing an interface instead?

**Comment by ibzib** [comment]
> I currently have streaming context as a subclass of the original spark translation context, but thinking some more I'm not sure if that totally works, and it might be better to have them both implementing an interface instead?

I don't see any problems currently. If we do find/introduce any inconsistencies or confusing behaviors, we can change it later.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkTranslationContext.java [hunk 1]
```diff
@@ -35,15 +35,15 @@
  * compute them after translation.
  */
 public class SparkTranslationContext {
-  private final JavaSparkContext jsc;
+  protected final JavaSparkContext jsc;
```
[the remark this answers is not in the document]
Comment (annaqin418) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkTranslationContext.java:38]: You're right, they shouldn't be, I got confused with the inheritance

**Comment by ibzib** [comment]
Please fix compilation errors:

```
> Task :runners:spark:compileJava FAILED
/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_Commit/src/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:52: warning: [MissingOverride] knownUrns implements method in SparkPortablePipelineTranslator; expected @Override
  public Set<String> knownUrns() {
                     ^
    (see http://errorprone.info/bugpattern/MissingOverride)
  Did you mean '@Override public Set<String> knownUrns() {'?
/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_Commit/src/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:77: warning: [MissingOverride] translate implements method in SparkPortablePipelineTranslator; expected @Override
  public void translate(
              ^
    (see http://errorprone.info/bugpattern/MissingOverride)
  Did you mean '@Override public void translate('?
/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_Commit/src/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:126: warning: [MissingOverride] createTranslationContext implements method in SparkPortablePipelineTranslator; expected @Override
  public SparkStreamingTranslationContext createTranslationContext(
                                          ^
    (see http://errorprone.info/bugpattern/MissingOverride)
  Did you mean '@Override public SparkStreamingTranslationContext createTranslationContext('?
```

File: runners/spark/job-server/build.gradle [hunk 2]
```diff
@@ -82,13 +82,21 @@ runShadow {
     jvmArgs += ["-Dorg.slf4j.simpleLogger.defaultLogLevel=${project.property('logLevel')}"]
 }
 
-def portableValidatesRunnerTask(String name) {
+def portableValidatesRunnerTask(String name, Boolean streaming) {
+  def pipelineOptions = [
+          // Limit resource consumption via parallelism
+          "--parallelism=2",
```
Comment (ibzib) [inline runners/spark/job-server/build.gradle:88]: This option only applies to Flink, it has no meaning in Spark. https://github.com/apache/beam/blob/103facbf80c2e4786e60367161ec59352196d04a/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java#L74

**Comment by annaqin418** [comment]
R: @robertwb 
This PR is still a while from being ready, but just wanted to tag you as well in my progress.

**Comment by aromanenko-dev** [comment]
@annaqin418 Is this PR WIP?

**Comment by ibzib** [comment]
Run Spotless PreCommit

**Comment by ibzib** [comment]
Java precommit failures are unrelated (see BEAM-7865)

**Comment by ibzib** [comment]
run seed job

**Comment by annaqin418** [comment]
R: @ibzib @robertwb 
I still need to document all of the limitations (e.g. side inputs, state/timers), but this code is ready for review

**Comment by ibzib** [comment]
Run Java Spark PortableValidatesRunner Streaming

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+
+  private static String getOutputId(PTransformNode transformNode) {
+    return Iterables.getOnlyElement(transformNode.getTransform().getOutputsMap().values());
+  }
+
+  private static String getExecutableStageIntermediateId(PTransformNode transformNode) {
+    return transformNode.getId();
+  }
+
+  private static <T> WindowedValue.WindowedValueCoder<T> getWindowedValueCoder(
```
Comment (ibzib) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:372]: We should move these common methods into a shared class like [`PipelineTranslatorUtils`](https://github.com/apache/beam/blob/master/runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/translation/PipelineTranslatorUtils.java).

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingTranslationContext.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+import org.apache.beam.runners.fnexecution.provisioning.JobInfo;
+import org.apache.beam.runners.spark.SparkPipelineOptions;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.streaming.Duration;
+import org.apache.spark.streaming.api.java.JavaStreamingContext;
+import org.joda.time.Instant;
+
+/**
+ * Translation context used to lazily store Spark data sets during portable pipeline translation and
```
Comment (ibzib) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingTranslationContext.java:28]: Nit: this comment is technically accurate, but it should mention streaming.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingTranslationContext.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  public SparkStreamingTranslationContext(
+      JavaSparkContext jsc, SparkPipelineOptions options, JobInfo jobInfo) {
+    super(jsc, options, jobInfo);
+    Duration batchDuration = new Duration(options.getBatchIntervalMillis());
+    this.streamingContext = new JavaStreamingContext(jsc, batchDuration);
+    this.firstTimestamp = new Instant();
+  }
+
+  public JavaStreamingContext getStreamingContext() {
```
Comment (ibzib) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingTranslationContext.java:43]: Make these methods package-private.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingTranslationContext.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.spark.streaming.api.java.JavaStreamingContext;
+import org.joda.time.Instant;
+
+/**
+ * Translation context used to lazily store Spark data sets during portable pipeline translation and
+ * compute them after translation.
+ */
+public class SparkStreamingTranslationContext extends SparkTranslationContext {
+  private final JavaStreamingContext streamingContext;
+  private final Instant firstTimestamp;
```
Comment (ibzib) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingTranslationContext.java:33]: Nit: `firstTimestamp` sounds a little strange here, since it's the only timestamp around. Maybe `initialTimestamp` or something?

File: runners/spark/src/test/java/org/apache/beam/runners/spark/SparkRunnerRegistrarTest.java [hunk 7]
```diff
@@ -37,7 +37,10 @@
   @Test
   public void testOptions() {
     assertEquals(
-        ImmutableList.of(SparkPipelineOptions.class, SparkStructuredStreamingPipelineOptions.class),
+        ImmutableList.of(
```
Comment (ibzib) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/SparkRunnerRegistrarTest.java:40]: Maybe we don't have to do it in this PR, but I'd be in favor of removing this test. It's kind of pointless to test a constant.

File: runners/spark/job-server/build.gradle [hunk 8]
```diff
@@ -82,53 +82,116 @@ runShadow {
...
+            testFilter: {
+              // TODO(BEAM-10094)
+              excludeTestsMatching 'org.apache.beam.sdk.transforms.FlattenTest.testFlattenWithDifferentInputAndOutputCoders2'
+              // Tests to exclude when streaming
+              excludeTestsMatching 'org.apache.beam.sdk.transforms.CombineTest'
+              excludeTestsMatching 'org.apache.beam.sdk.transforms.CombineFnsTest'
+              excludeTestsMatching 'org.apache.beam.sdk.transforms.ParDoTest$MultipleInputsAndOutputTests'
+              excludeTestsMatching 'org.apache.beam.sdk.transforms.ParDoTest$StateCoderInferenceTests'
+              excludeTestsMatching 'org.apache.beam.sdk.transforms.ParDoTest$StateTests'
+              excludeTestsMatching 'org.apache.beam.sdk.transforms.ParDoTest$TimerTests'
```
Comment (ibzib) [inline runners/spark/job-server/build.gradle:138]: If possible, exclude test categories instead of test classes.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPortableStreamingPipelineOptions.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Spark runner {@link PipelineOptions} handles Spark execution-related configurations, such as the
+ * master address, batch-interval, and other user-related knobs.
+ */
+@Experimental
+public interface SparkPortableStreamingPipelineOptions
+    extends SparkPipelineOptions, PortablePipelineOptions, PipelineOptions {
+  @Description("Timeout for testing Spark portable streaming, in milliseconds.")
+  @Default.Long(-1L)
```
[the remark this answers is not in the document]
Comment (annaqin418) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPortableStreamingPipelineOptions.java:34]: This is a good question, I couldn't find documentation on the proper way to have no timeout. Typically it seems that you would just use two entirely separate methods, awaitTermination() and awaitTerminationOrTimeout(). -1L simply appeared to give the desired behavior...

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingTranslationContext.java (hunk 5 above)
Comment (annaqin418) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingTranslationContext.java:43]: I'm not sure I can do so, since the SparkPipelineRunner calls this method but is outside of the translation package?

Comment (ibzib) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingTranslationContext.java:43]: Ah okay. Disregard then.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPortableStreamingPipelineOptions.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ibzib) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPortableStreamingPipelineOptions.java:34]: It's not really documented anywhere, but I found it in the Spark source. Looks like `awaitTermination()` is indeed equivalent to `awaitTerminationOrTimeout(-1)`.

https://github.com/apache/spark/blob/b03761e3303e979999d4faa5cf4d1719a82e06cb/streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala#L637

https://github.com/apache/spark/blob/b03761e3303e979999d4faa5cf4d1719a82e06cb/streaming/src/main/scala/org/apache/spark/streaming/ContextWaiter.scala#L58

File: runners/spark/job-server/build.gradle [hunk 10]
```diff
@@ -82,20 +82,60 @@ runShadow {
...
+      excludeCategories 'org.apache.beam.sdk.testing.UsesOnWindowExpiration'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesTestStream'
+      // SplittableDoFnTests
+      excludeCategories 'org.apache.beam.sdk.testing.UsesBoundedSplittableParDo'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesSplittableParDoWithWindowedSideInputs'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesUnboundedSplittableParDo'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesStrictTimerOrdering'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesBundleFinalizer'
+      // Currently unsupported in portable streaming
+      excludeCategories 'org.apache.beam.sdk.testing.UsesSideInputs'
```
Comment (ibzib) [inline runners/spark/job-server/build.gradle:117]: Link JIRAs for side input, state, and timer support.

**Comment by annaqin418** [comment]
run seed job

File: runners/spark/job-server/build.gradle [hunk 11]
```diff
@@ -82,20 +82,68 @@ runShadow {
...
+      excludeCategories 'org.apache.beam.sdk.testing.UsesGaugeMetrics'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesParDoLifecycle'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesMapState'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesSetState'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesOrderedListState'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesTimerMap'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesKeyInParDo'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesOnWindowExpiration'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesTestStream'
+      // SplittableDoFnTests
```
Comment (ibzib) [inline runners/spark/job-server/build.gradle:110]: This should've been there before, but the jira for this (BEAM-7222) should be linked.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkExecutableStageFunction.java [hunk 12]
```diff
@@ -118,6 +118,12 @@
 
   @Override
   public Iterator<RawUnionValue> call(Iterator<WindowedValue<InputT>> inputs) throws Exception {
+    // Do not call processElements if there are no inputs
+    // Otherwise, this may cause validation errors (e.g. ParDoTest)
+    if (!inputs.hasNext()) {
+      return new ConcurrentLinkedQueue<RawUnionValue>().iterator();
```
Comment (ibzib) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkExecutableStageFunction.java:124]: `Collections.emptyIterator()` has the same behavior, but is more concise.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkPortablePipelineTranslator.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.Set;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.runners.fnexecution.provisioning.JobInfo;
+import org.apache.beam.runners.spark.SparkPipelineOptions;
+import org.apache.spark.api.java.JavaSparkContext;
+
+/**
+ * Interface for portable Spark translators. This allows for a uniform invocation pattern for
+ * pipeline translation between streaming and portable runners.
```
Comment (ibzib) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkPortablePipelineTranslator.java:28]: "streaming and portable" should be "streaming and batch."

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/SparkExecutableStageFunctionTest.java [hunk 14]
```diff
@@ -101,7 +104,9 @@ public void setUpMocks() throws Exception {
   public void sdkErrorsSurfaceOnClose() throws Exception {
     SparkExecutableStageFunction<Integer, ?> function = getFunction(Collections.emptyMap());
     doThrow(new Exception()).when(remoteBundle).close();
-    function.call(Collections.emptyIterator());
+    List<WindowedValue<Integer>> dummyList = new ArrayList<>();
```
Comment (ibzib) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/SparkExecutableStageFunctionTest.java:107]: Nit: "dummy" sounds a lot like mocks or stubs, which are technical terms that describe the implementation of test objects. This list isn't a "dummy" in that sense; it's a real list, albeit one constructed for test purposes.

tl;dr rename this `inputs`.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+  }
+
+  private static void translateImpulse(
+      PTransformNode transformNode,
+      RunnerApi.Pipeline pipeline,
+      SparkStreamingTranslationContext context) {
+
+    TimestampedValue<byte[]> tsValue = TimestampedValue.atMinimumTimestamp(new byte[0]);
+    Iterable<TimestampedValue<byte[]>> timestampedValues = Collections.singletonList(tsValue);
+    Iterable<WindowedValue<byte[]>> windowedValues =
```
Comment (robertwb) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:151]: It seems it'd be easier construct this with

```
Collections.singletonList(
    WindowedValue.of(...));
```

rather than use an intermediate timestamped value list.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+                        PaneInfo.NO_FIRING))
+            .collect(Collectors.toList());
+
+    ByteArrayCoder coder = ByteArrayCoder.of();
+    WindowedValue.FullWindowedValueCoder<byte[]> windowCoder =
+        WindowedValue.FullWindowedValueCoder.of(coder, GlobalWindow.Coder.INSTANCE);
+    JavaRDD<WindowedValue<byte[]>> emptyByteArrayRDD =
+        context
+            .getSparkContext()
+            .parallelize(CoderHelpers.toByteArrays(windowedValues, windowCoder))
```
Comment (robertwb) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:168]: You could even defer the singleton iterable wrapping to here and use `windowCoder.encode(...)` rather than use `CoderHelpers.toByteArrays`.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+            .map(CoderHelpers.fromByteFunction(windowCoder));
+
+    Queue<JavaRDD<WindowedValue<byte[]>>> rddQueue = new LinkedBlockingQueue<>();
+    rddQueue.offer(emptyByteArrayRDD);
+    JavaInputDStream<WindowedValue<byte[]>> emptyStream =
```
Comment (robertwb) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:173]: This stream isn't empty, is it?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+        new UnboundedDataset<>(
+            emptyStream, Collections.singletonList(emptyStream.inputDStream().id()));
+
+    // Add watermark to holder and advance to infinity to ensure future watermarks can be updated
+    GlobalWatermarkHolder.SparkWatermarks sparkWatermark =
+        new GlobalWatermarkHolder.SparkWatermarks(
+            GlobalWindow.INSTANCE.maxTimestamp(),
+            BoundedWindow.TIMESTAMP_MAX_VALUE,
+            context.getFirstTimestamp());
+    GlobalWatermarkHolder.add(output.getStreamSources().get(0), sparkWatermark);
```
Comment (robertwb) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:186]: I have to admit I'm a bit fuzzy on how this holds back (or doesn't) the watermark.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+    UnboundedDataset<KV<K, V>> inputDataset =
+        (UnboundedDataset<KV<K, V>>) context.popDataset(inputId);
+    List<Integer> streamSources = inputDataset.getStreamSources();
+    JavaDStream<WindowedValue<KV<K, V>>> dStream = inputDataset.getDStream();
+    WindowedValue.WindowedValueCoder<KV<K, V>> inputCoder =
+        getWindowedValueCoder(inputId, components);
+    KvCoder<K, V> inputKvCoder = (KvCoder<K, V>) inputCoder.getValueCoder();
+    Coder<K> inputKeyCoder = inputKvCoder.getKeyCoder();
+    Coder<V> inputValueCoder = inputKvCoder.getValueCoder();
+    final WindowingStrategy windowingStrategy = getWindowingStrategy(inputId, components);
```
Comment (robertwb) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:206]: Be consistent on what's declared final vs. what's not.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+              transformNode.getTransform().getSpec().getPayload());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    String inputPCollectionId = stagePayload.getInput();
+    UnboundedDataset<InputT> inputDataset =
+        (UnboundedDataset<InputT>) context.popDataset(inputPCollectionId);
+    List<Integer> streamSources = inputDataset.getStreamSources();
+    JavaDStream<WindowedValue<InputT>> inputDStream = inputDataset.getDStream();
```
Comment (robertwb) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:242]: Nit: I'm noticing the pattern of assigning a huge number of locals at the top of a function and then only using them once below. Often (not always) the code would be more concise if the declarations were used inline (e.g. use `inputDataset.getDStream()` rather than assign to an intermediate `inputDStream`).

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java [hunk 21]
```diff
@@ -0,0 +1,366 @@
...
+    Map<String, String> outputs = transformNode.getTransform().getOutputsMap();
+    BiMap<String, Integer> outputMap = createOutputMap(outputs.values());
+
+    RunnerApi.Components components = pipeline.getComponents();
+    Coder windowCoder =
+        getWindowingStrategy(inputPCollectionId, components).getWindowFn().windowCoder();
+
+    // TODO (BEAM-10712): handle side inputs.
```
Comment (robertwb) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:250]: Trow a clear exception here if the executable stage declares itself as having side inputs.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java [hunk 22]
```diff
@@ -0,0 +1,366 @@
...
+    context.popDataset(intermediateId);
+
+    for (String outputId : outputs.values()) {
+      JavaDStream<WindowedValue<OutputT>> outStream =
+          staged.flatMap(new SparkExecutableStageExtractionFunction<>(outputMap.get(outputId)));
+      context.pushDataset(outputId, new UnboundedDataset<>(outStream, streamSources));
+    }
+
+    if (outputs.isEmpty()) {
+      // Add sink to ensure all outputs are computed
```
Comment (robertwb) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:297]: In this case there are no outputs (per the above line). Perhaps just say something to the effect that we want to ensure the stage is executed.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java [hunk 23]
```diff
@@ -0,0 +1,366 @@
...
+        Dataset dataset = context.popDataset(inputId);
+        if (dataset instanceof UnboundedDataset) {
+          UnboundedDataset<T> unboundedDataset = (UnboundedDataset<T>) dataset;
+          streamSources.addAll(unboundedDataset.getStreamSources());
+          dStreams.add(unboundedDataset.getDStream());
+        } else {
+          // create a single RDD stream.
+          Queue<JavaRDD<WindowedValue<T>>> q = new LinkedBlockingQueue<>();
+          q.offer(((BoundedDataset) dataset).getRDD());
+          // TODO: this is not recoverable from checkpoint!
```
Comment (robertwb) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkStreamingPortablePipelineTranslator.java:330]: File and reference a JIRA.
